hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ConditionalWriteLockProvider.java

```diff
@@ -0,0 +1,581 @@
+    }
+    this.currentLockObj = lockObj;
+  }
+
+  /**
+   * Default constructor for ConditionalWriteLockProvider, required by LockManager
+   * to instantiate it using reflection.
+   * 
+   * @param lockConfiguration The lock configuration, should be transformable into
+   *                          ConditionalWriteLockConfig
+   * @param conf              Storage config, ignored.
+   */
+  public ConditionalWriteLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    ConditionalWriteLockConfig config = new ConditionalWriteLockConfig.Builder()
+        .fromProperties(lockConfiguration.getConfig()).build();
+    heartbeatIntervalMs = config.getHeartbeatPollMs();
+    lockValidityMs = config.getLockValidityTimeoutMs();
+
+    // Determine if the provided locks location is relative.
+    String configuredLocksLocation = config.getLocksLocation();
+
+    // If using base path, recalculate the locks location as .hoodie/.locks
+    String locksLocation = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? String.format("%s%s%s", config.getHudiTableBasePath(), StoragePath.SEPARATOR, LOCKS_FOLDER_NAME)
+        : configuredLocksLocation;
+
+    URI uri = parseURI(locksLocation);
+    bucketName = uri.getHost(); // For most schemes, the bucket/container is the host.
+    String folderName = uri.getPath(); // Path after the bucket/container.
+
+    String fileName = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? DEFAULT_TABLE_LOCK_FILE_NAME
+        : slugifyLockFolderFromBasePath(config.getHudiTableBasePath());
+
+    lockFilePath = buildLockObjectPath(folderName, fileName);
+    ownerId = UUID.randomUUID().toString();
+    this.logger = DEFAULT_LOGGER;
+    this.heartbeatManager = new LockProviderHeartbeatManager(
+        ownerId,
+        heartbeatIntervalMs,
+        this::renewLock);
+
+    try {
+      this.lockService = (StorageLock) ReflectionUtils.loadClass(
+          getLockServiceClassName(uri.getScheme()),
+          new Class<?>[] { String.class, String.class, String.class, Properties.class },
+          new Object[] { ownerId, bucketName, lockFilePath, lockConfiguration.getConfig() });
+    } catch (Throwable e) {
+      throw new HoodieLockException("Failed to load and initialize StorageLock", e);
+    }
+
+    logger.info("Instantiated new Conditional Write LP, owner: {}, lockfilePath: {}", ownerId, lockFilePath);
+  }
+
+  private URI parseURI(String location) {
+    try {
+      return new URI(location);
+    } catch (URISyntaxException e) {
+      throw new HoodieLockException("Unable to parse locks location as a URI: " + location, e);
+    }
+  }
+
+  private static @NotNull String getLockServiceClassName(String scheme) {
+    if (SUPPORTED_LOCK_SERVICE_IMPLEMENTATIONS.containsKey(scheme)
+        && StorageSchemes.isConditionalWriteSupported(scheme)) {
+      return SUPPORTED_LOCK_SERVICE_IMPLEMENTATIONS.get(scheme);
```

Comment:
There are two ways to store the class name:
- Store the mapping in this class
- Let `StorageSchemes` return the lock service class name so this class does not have any storage-specific information.

I think either way is fine for now because only two file systems are supported.  If you have time you can switch to the second way.

Reply:
This would require a dependency on hudi-client-common in hudi-io. Is that ok?

Or maybe we move that interface to hudi-io?

Reply:
Discussed offline that we use a String to store the class name.